[PyTorch][Dist] Trigger pre/post hooks of output function nodes under distributed autograd

hczhu opened the pull request:
# Goals
Do the following things during a distributed backward pass.
1. Accumulate the gradient of a variable to RPC context once the gradient is ready instead of at the very end of the backward pass.
2. Run post/pre hooks installed in`AccumulateGrad` nodes once the gradient is ready for the variable. Currently, the hooks in `AccumulateGrad` are not executed just because the function `AccumulateGrad` itself is not even evaluated by the local engine.
3. Make it extensible to support post hooks installed by DDP's reducer.


# Introduce GradCapturePreHook

## Why do we need this?

### Root issue:

* dist engine uses the autograd.grad-like API on the vanilla engine and then in the Future callback populates the context with the gradients. This is a bad emulation of the .backward() call on the vanilla engine.

### Practical issue: 

* The leaf’s hook are not called (because associated with the AccumulateGrad that is not call in the autograd.grad-like API). Modules like DDP rely on these hooks.
* The Future is marked as completed before the context is actually populated with the grads leading to unexpected behavior on the user side.
* The Future callback is only called at the complete end of the backward and so too late for DDP if they want to overlap compute/transfert.

### Proposed solution:

* Provide hooks in the autograd.grad-like API that will allow the distributed engine to populate the context and call the hooks to better emulate the .backward call.

## Who can install a grad capture pre-hook?

This will be an internal hook at C++ level and it won’t be exposed to PyThon code. Only call-sites directly interacting with the local engine can install such hooks.

## Signature
The returned `grad` will be captured.
```
virtual const torch::Tensor& grad operator()(const torch::Tensor& grads) = 0;
```

## Where are hooks installed?

Grad capture pre-hooks are install in GraphTask::ExecInfo::Capture. ExecInfo is per node. Every backward run will have its own GraphTask instance.

## When/How will hooks be called?

When the local engine captures the grads for a node, all grad capture pre hooks are called one by one in the order they are added. The output grads of the hooks will replace the original grads.
The output of the last hook will be used for grad capturing.

# Test plan

All existing tests should pass.
```
python setup.py develop

python test/distributed/rpc/test_dist_autograd_spawn.py DistAutogradTestWithSpawn.test_post_hooks

make test_dist_autograd && bin/test_dist_autograd

```

albanD commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -173,6 +222,11 @@ void DistEngine::computeDependencies(
+    // Mark all output nodes as needed.
+    for (const auto& outputEdge : outputEdges) {
+      const auto func = outputEdge.function.get();
+      graphTask->exec_info_[func].needed_ = true;
```
All the exec_info_ can be removed now. Because you want the whole graph to be processed.

albanD commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -173,6 +222,11 @@ void DistEngine::computeDependencies(
+    // Mark all output nodes as needed.
+    for (const auto& outputEdge : outputEdges) {
+      const auto func = outputEdge.function.get();
+      graphTask->exec_info_[func].needed_ = true;
```
All the exec_info_ can be removed now. Because you want the whole graph to be processed.

hczhu replied:
Will do that.

@mrshenli @pritamdamania  could you also help confirm that `exec_info_` can be deleted?

hczhu replied:
@albanD I think I misunderstood you. What does "exec_info_ can be removed" mean? `exec_info_` is a field of `GraphTask` and is also used by the local engine to determine whether to apply a function.

albanD replied:
You can double check [this comment in the code](https://github.com/pytorch/pytorch/blob/20b18a58f19a9b5907697e4b1ffe075d3fc2ec52/torch/csrc/autograd/engine.h#L71-L77). But the exec_info_ field has a special semantic when its left empty.

Namely:
- exec_info_ == nullptr corresponds to a `loss.backward()` in python: the whole graph is processed with no exception.
- exec_info_ != nullptr corresponds to `autograd.grad(loss, inputs)` in python: only the graph between loss and inputs is processed.

Now that you change the graph to remove the AccumulateGrad, you actually want to run the engine just like `.backward()` as you want everything to run.

hczhu replied:
How can `exec_info_` be `nullptr`? It's of type `std:: unordered_map` instead of a pointer. I guess you meant the map is empty.

I'm not sure we want everything to run. The signature of distributed backward is `backward(context_id: int, roots: List[Tensor], retain_graph = False)`. Only nodes of the autograd graph reachable from `roots` need to be run.
@mrshenli @pritamdamania87

albanD replied:
We always start at the roots, so everything that is not reachable from the roots won't run.
The signature is the same as the regular engine backward(). Which maps to an empty exec_info_.

Yes sorry, I meant `.empty()` not nullptr, you can see the check [here](https://github.com/pytorch/pytorch/blob/cd9d9a2235567eb76743a5d863e0f2d4074d8956/torch/csrc/autograd/engine.cpp#L559).

albanD replied:
Yes sorry I meant empty, not nullptr.

The backward is the same as the regular engine's backward that runs without exec_info_.
The current implementation uses exec_info_ because it needs to bypass the `AccumulateGrad` Nodes that should not run. In this case, since you remove them, this is not needed anymore.

albanD commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -131,13 +139,54 @@ void DistEngine::computeDependencies(
+                  // to 'distAccumulateGradFn'.
+                  std::move(*accumulateGradFn),
+                  autogradContext);
+              distAccumulateGradFn->replace_grad_accumulator();
```
Why is this needed?
Changing the `next_edge` in the graph below means that we won't even try to use the grad_accumulator.

hczhu replied:
All output tensor/variable keeps a weak_ptr to its `AccumulateGrad`. Would it be a problem if the `AccumulateGrad` functions in autograd graph are inconsistent with the ones in tensors/variables?

albanD replied:
The `AccumulateGrad` Node is the one that populates the `.grad` field of a variable.
You could modify the grad_accumulator before the graph is created. But that sounds really weird as you have no way to know about it.
So what you want is just to replace the node currently in the graph to avoid accumulating in the `.grad` field.

albanD commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -64,6 +65,8 @@ void DistEngine::validateRootsAndRetrieveEdges(
       rootEdges, grads, [](const std::string& msg) { return msg; });
 }
 
+// NB: this function modifies the autograd graph. More specifically it replaces
```
Note that since the graph could be used in other places, and in particular it could be used with the vanilla engine, you want to be careful here.
You should either forbid `retain_graph` to make sure no one will run the graph again or you should make this properly work in this case.

hczhu replied:
I think it should work for `retain_graph = true`. I'll test that.

albanD replied:
In particular if you do mix the vanilla engine and the distributed one.

albanD commented on torch/csrc/distributed/autograd/functions/dist_accumulate_grad.cpp:
```diff
@@ -0,0 +1,49 @@
+    : Node(accumulateGrad.sequence_nr()),
+      variable_(std::move(accumulateGrad.variable)),
+      autogradContext_(std::move(autogradContext)) {
+  move_from(std::move(accumulateGrad));
```
I don't think you want to steal stuff from the AccumulateGrad. The AccumulateGrad node can outlive this single backward call and might be used again.

hczhu replied:
The `AccumulateGrad` node is not a part of the autograd graph. How can others refer to it?
(though I know DDP reducers, which create post hooks, keep references to them)

https://github.com/pytorch/pytorch/blob/15a7b9cf0ab4674e798d735042cdcc9184b66e8a/torch/csrc/distributed/c10d/reducer.cpp#L195-L207

albanD replied:
It is actually in the graph. This is why you get it by looking into `next_functions`. Being in the graph is actually the only thing that keeps it alive as, as you said, the variable itself only has a weak_ref to it.

hczhu replied:
`DistAccumulateGrad` is a replacement of `AccumulateGrad` in the graph. Here it has to steal metadata, most of which are in `unique_ptr`, form `AccumulateGrad`. I know `AccumulateGrad` nodes can outlive the graph. One example is DDP reducer keeps `shared_ptr` to them.

albanD replied:
But what happens if someone does:
```
inp = torch.rand(10, requires_grad=True)
loss = model(inp)
with dist_autograd.context() as context_id:
  dist_autograd.backward([loss], retain_graph=True)
loss.backward()
```

I don't claim I have a use case for this. But if we don't support this, it should be explicitly checked and should raise a nice error.

mrshenli replied:
For existing DDP, calling backward twice would crash, as it requires a forward pass to mark `expect_autograd_hooks_` to `true`.

hczhu replied:
> But what happens if someone does:
> 
> ```
> inp = torch.rand(10, requires_grad=True)
> loss = model(inp)
> with dist_autograd.context() as context_id:
>   dist_autograd.backward([loss], retain_graph=True)
> loss.backward()
> ```
> 
> I don't claim I have a use case for this. But if we don't support this, it should be explicitly checked and should raise a nice error.

I'll make this one a test case.

albanD commented on torch/csrc/distributed/autograd/functions/dist_accumulate_grad.cpp:
```diff
@@ -0,0 +1,49 @@
+    return; 
+  }
+  // Replace the variable's reference to AccumulateGrad.
+  torch::autograd::impl::set_grad_accumulator(
```
As mentioned above, you don't want to replace the original one. Just change the current graph you were given.

albanD reviewed (changes requested):
The gradient accumulation logic looks much better.

My main worry is that the graph can actually outlive this call to backward and should be restored.
The leaf also will most likely outlive this call and will be re-used so it's grad accumulator should not be changed.

albanD commented:
@hczhu  Small questions about the description:
- What do you mean by "Currently, function hooks are ignored." ?  Which hooks in particular?
- "Will any component use the replaced AccumulateGrad nodes during a backward pass?" The particular use cases I have in mind are:
  - What happens if I run backward() on the same graph from two threads from two different distributed contexts?
  - What happens if I run backward() on the same graph from two thread, one using distributed autograd and one using regular autograd?

mrshenli commented on torch/csrc/autograd/function.h:
```diff
@@ -84,7 +84,7 @@ TORCH_API void deleteNode(Node* function);
 /// guarantees* w.r.t. the ordering of `C` relative to `A` or `B`.
 ///~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-struct TORCH_API Node : std::enable_shared_from_this<Node> {
+struct TORCH_API Node : public std::enable_shared_from_this<Node> {
```
Why do we need this change? The default should be public for struct, no?

hczhu replied:
You are right. I'll revert it.

mrshenli commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -131,13 +139,54 @@ void DistEngine::computeDependencies(
+            } else if (
+                auto accumulateGradFn = dynamic_cast<AccumulateGrad*>(nextFn)) {
+              TORCH_INTERNAL_ASSERT(
+                  !accumulateGradFn->variable.defined() ||
```
Curious, when would the variable be not defined?

hczhu replied:
Deleted this. That may be caused by my incorrect graph modification.

mrshenli commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -131,13 +139,54 @@ void DistEngine::computeDependencies(
+                          torch::autograd::impl::try_get_grad_accumulator(
+                              accumulateGradFn->variable)
+                              .get(),
+                  "The AccumulateGrad point in the variable should be the "
```
s/point/node/g?

hczhu replied:
deleted

mrshenli commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -99,11 +102,16 @@ void DistEngine::computeDependencies(
-    for (const auto& edge : fn->next_edges()) {
+    for (int index = 0; index < fn->num_outputs(); ++index) {
+      const auto& edge = fn->next_edge(index);
       if (auto nextFn = edge.function.get()) {
```
Why do we need a raw pointer here? Is it just to avoid the overhead of reference counting? IIUC, `edge.function` is a shared_ptr.

hczhu replied:
Legacy code. The reasons seem
1. The shared_ptr may be nullptr. Therefore, an `if` is here.
2. All containers in this function, like `queue` and `dependencies `, use raw pointers. It's more convenient to have a raw pointer.

mrshenli commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -131,13 +139,54 @@ void DistEngine::computeDependencies(
+                  // The pointer to 'AccumulateGrad' in
+                  // 'accumulateGradFn->variable' will be replaced by a pointer
+                  // to 'distAccumulateGradFn'.
+                  std::move(*accumulateGradFn),
```
Why do we need to move the `AccumulateGrad` node here? Can we just keep a copy the `accumulateGradFn ` `shared_ptr` (assuming we can get shared_ptr from `edge.function`)? This will keep the existing `AccumulateGrad` node intact?

mrshenli commented:
Looks like we need the move-restore logic because `Node::post_hooks()` is not virtual? As Alban asked above: 

> What happens if I run backward() on the same graph from two threads from two different distributed contexts?

It seems the move-restore strategy would not work here, as the second thread would see an empty `AccumulateGrad` node?

@albanD @ezyang how bad is it if we make `Node::post_hooks()` virtual? That could make this effort easier and more robust.

gqchen commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -173,6 +222,11 @@ void DistEngine::computeDependencies(
       graphTask->exec_info_[recvBackwardEdge.function.get()].needed_ = true;
     }
+    // Mark all output nodes as needed.
+    for (const auto& outputEdge : outputEdges) {
```
Not sure why we need to explicitly mark those outputEdge functions. Would those functions already be marked properly in "graphTask->init_to_execute" in line 219?

hczhu replied:
Deleted according to @albanD 's suggestion.

gqchen commented on torch/csrc/distributed/autograd/functions/dist_accumulate_grad.cpp:
```diff
@@ -0,0 +1,49 @@
+    : Node(accumulateGrad.sequence_nr()),
+      variable_(std::move(accumulateGrad.variable)),
+      autogradContext_(std::move(autogradContext)) {
+  move_from(std::move(accumulateGrad));
```
can we maintain a pointer to AccumulateGrad node in DistAccumulateGrad, and use those contents in AccumulateGrad through the pointer?

hczhu replied:
Delegating to `AccumulateGrad` should be better, but `Node:: post_hooks()` needs to be virtual (it's not right now). Making it(together with other Node functions) virtual may cause performance issue.

hczhu commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -64,6 +65,8 @@ void DistEngine::validateRootsAndRetrieveEdges(
 }
 
+// NB: this function modifies the autograd graph. More specifically it replaces
+// all AccumulateGrad nodes by DistAccumulateGrad.
```
> it seems to be confusing that we are replacing nodes in a function called computeDependencies. It might be better to write a new function to do node replacement. We saved some time by avoiding one more traversal. But now the code is much harder to understand and maintain.

I thought about another using another function to traverse the graph again. My consideration is for performance.

hczhu replied:
@albanD @mrshenli What do you guys think about another graph traversal?

albanD replied:
The vanilla engine does two when using autograd.grad.
The good thing about splitting them is that you will be able to use the `compute_dependencies` function from the original engine and that reduces code duplication.

hczhu commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -131,13 +139,54 @@ void DistEngine::computeDependencies(
+              distAccumulateGradFn->replace_grad_accumulator();
+              accumulate_grad_replacements[nextFn] = distAccumulateGradFn;
+              fn->next_edge(index).function = std::move(distAccumulateGradFn);
+              TORCH_INTERNAL_ASSERT(edge == fn->next_edge(index));
```
Deleted.

hczhu commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -131,13 +139,54 @@ void DistEngine::computeDependencies(
+        "Replaced AccumulateGrad nodes should be in dependencies.");
+    dependencies[distAccumulateGrad.get()] = itr->second;
+    dependencies.erase(itr);
+  }
```
Yes, I piggybacked the graph modification on compute dependencies just to avoid another graph traversing.

hczhu replied:
Deleted already.

hczhu commented:
> @hczhu Small questions about the description:
> 
> * What do you mean by "Currently, function hooks are ignored." ?  Which hooks in particular?

Updated the summary.


> * "Will any component use the replaced AccumulateGrad nodes during a backward pass?" The particular use cases I have in mind are:
>   
>   * What happens if I run backward() on the same graph from two threads from two different distributed contexts?
>   * What happens if I run backward() on the same graph from two thread, one using distributed autograd and one using regular autograd?

Good thoughts. I'll make those 2 situations test cases.

hczhu commented on test/distributed/test_dist_accumulate_grad.py:
```diff
@@ -0,0 +1,143 @@
+            tensor_to_grad = dist_autograd.get_gradients(context_id)
+            gLogger.info(f"Got tensor to grad map: {tensor_to_grad}")
+            
+    def test_single_rpc(self):
```
Will add more test cases here.

hczhu commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -33,6 +35,21 @@ DistEngine& DistEngine::getInstance() {
+        std::move(accumulateGradFn);
+    context.distAccumulateGradCtx.pop();
+  }
+  LOG(ERROR) << "Restoring the graph is done.";
```
Will remove all such logs.

albanD commented:
> It seems the move-restore strategy would not work here, as the second thread would see an empty AccumulateGrad node?

@mrshenli I agree that this might be a big shortcoming of this strategy...
The thing is that even the other strategy that would modify the AccumulateGrad would have the same issue as such Node runs on the same worker threads, whatever was the context in which  the backward was initiated.

> how bad is it if we make Node::post_hooks() virtual? That could make this effort easier and more robust.

This functions is called for every Node evaluation. So there can be a significant impact.
But the concurent backward problem above should be solved first I think.

hczhu commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -33,6 +35,21 @@ DistEngine& DistEngine::getInstance() {
+            oneEdge.from->next_edge(oneEdge.edgeIndex).function.get()));
+    auto accumulateGradFn = oneEdge.to->restoreAccumulateGrad();
+    oneEdge.from->next_edge(oneEdge.edgeIndex).function =
+        std::move(accumulateGradFn);
```
There is a mistake here. I'll rewrite graph restoration shortly.

hczhu commented:
> Looks like we need the move-restore logic because `Node::post_hooks()` is not virtual? As Alban asked above:
> 
> > What happens if I run backward() on the same graph from two threads from two different distributed contexts?
> 
> It seems the move-restore strategy would not work here, as the second thread would see an empty `AccumulateGrad` node?
> 

Is distributed autograd designed to support that? The typical case seems multiple trainers remote call forward/backward through RPC and each thread has its own autograd graph sharing tensors/variables.

@mrshenli @albanD

hczhu commented:
> > Looks like we need the move-restore logic because `Node::post_hooks()` is not virtual? As Alban asked above:
> > > What happens if I run backward() on the same graph from two threads from two different distributed contexts?
> > 
> > 
> > It seems the move-restore strategy would not work here, as the second thread would see an empty `AccumulateGrad` node?

Just realized that under distributed engine, each forward pass (graph creation) has its own context id. Therefore, each backward pass will run on its own graph instance. @albanD The situation you mentioned can't happen by the current dist autograd design.

albanD commented:
> Just realized that under distributed engine, each forward pass (graph creation) has its own context id. Therefore, each backward pass will run on its own graph instance.

If my graph does not contain any Send/Recv nodes, then I don't have to run the forward inside the ctx right?

hczhu commented on torch/csrc/autograd/function.h:
```diff
@@ -343,6 +347,9 @@ struct TORCH_API Node : std::enable_shared_from_this<Node> {
   /// Calls `apply()`, but instruments it with tracing machinery.
   variable_list traced_apply(variable_list inputs);
 
+  // Move/copy all fields from 'rhs'.
```
> why not use the move constructor?

I wanted to explicit say the semantics of "moving metadata" and the moved-from instance can be reused later. Usually we don't reuse a moved-from instance. Our use case here is quite unique.

```
auto a = std::move(b);

// The common pattern is "never use b again"

// Our use case is 
// Restore b
b = std::move(a);
```
Using move ctor seems a bit anti-intuitive here.
@gqchen

hczhu commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -131,13 +139,54 @@ void DistEngine::computeDependencies(
+              // because the former accumulates grads to the variable's '.grad'
+              // without considering the context id. That may cause data race
+              // on the '.grad', since multiple context ids may have grads for
+              // the same '.grad'.
```
done.

hczhu commented:
@albanD @mrshenli 
Do the following examples cover all concurrent cases?

Example 1:
```Python
# forward pass has not RPC
loss = ...

# thread 1
loss.backward()

# thread 2
with dist_autograd.context() as context_id:
  dist_autograd.backward(context_id, [loss])

```
Example 2
```
# forward pass has not RPC
loss = ...

# thread 1
with dist_autograd.context() as context_id:
  dist_autograd.backward(context_id, [loss])

# thread 2
with dist_autograd.context() as context_id:
  dist_autograd.backward(context_id, [loss])

```

hczhu commented on torch/csrc/distributed/autograd/functions/dist_accumulate_grad.cpp:
```diff
@@ -0,0 +1,37 @@
+namespace torch::distributed::autograd {
+
+DistAccumulateGrad::DistAccumulateGrad(
+      std::shared_ptr<AccumulateGrad> accumulateGrad,
```
> [performance-unnecessary-value-param] warning: the parameter 'accumulateGrad' is copied for each invocation but only used as a const reference; consider making it a const reference

Don't quite understand this suggestion. The safe way to use `shared_ptr` is copying the value so that the underlying instance never gets freed.

ezyang commented:
> @albanD @ezyang how bad is it if we make Node::post_hooks() virtual? That could make this effort easier and more robust.

Actually, I'm going to disagree with Alban here; it probably doesn't matter if post hooks is virtual or not.

mrshenli commented:
> @albanD @mrshenli
> Do the following examples cover all concurrent cases?

I wonder why we would we use dist_autograd when there is no RPC in the forward pass? We probably should error out explicitly in this case and tell the application to use local autograd engine?

A 3rd example for smart mode + no_sync would be:

```python
# thread 1
with dist_autograd.context() as context_id:
  with ddp.no_sync():
    for _ in range(5):
      loss = ddp(input).sum()
      dist_autograd.backward(context_id, [loss])
  loss = ddp(input).sum()
  dist_autograd.backward(context_id, [loss])

# thread 2
with dist_autograd.context() as context_id:
  with ddp.no_sync():
    for _ in range(5):
      loss = ddp(input).sum()
      dist_autograd.backward(context_id, [loss])
  loss = ddp(input).sum()
  dist_autograd.backward(context_id, [loss])
```

mrshenli commented:
> Is distributed autograd designed to support that? The typical case seems multiple trainers remote call forward/backward through RPC and each thread has its own autograd graph sharing tensors/variables.
>
> @mrshenli @albanD

For the same variable in different threads, will their `AccumulateGrad` node be the same or will different GraphTasks have different `AccumulateGrad` node? I understand all other nodes in the graph should be different, but I am not sure about `AccumulateGrad`.

albanD commented:
If you do two different ops, you will get two different Nodes. So the graphs are distinct as you mention. But the AccumulateGrad is associated with the leaf. So if the leaf is used in multiple places, the same AccumulateGrad will be used.

hczhu commented:
> > @albanD @mrshenli
> > Do the following examples cover all concurrent cases?
> 
> I wonder why we would we use dist_autograd when there is no RPC in the forward pass? We probably should error out explicitly in this case and tell the application to use local autograd engine?
> 
> A 3rd example for smart mode + no_sync would be:
> 
> ```python
> # thread 1
> with dist_autograd.context() as context_id:
>   with ddp.no_sync():
>     for _ in range(5):
>       loss = ddp(input).sum()
>       dist_autograd.backward(context_id, [loss])
>   loss = ddp(input).sum()
>   dist_autograd.backward(context_id, [loss])
> 
> # thread 2
> with dist_autograd.context() as context_id:
>   with ddp.no_sync():
>     for _ in range(5):
>       loss = ddp(input).sum()
>       dist_autograd.backward(context_id, [loss])
>   loss = ddp(input).sum()
>   dist_autograd.backward(context_id, [loss])
> ```

@mrshenli Example 3 should just work, because each thread has its own graph instance created during the forward pass.

hczhu commented:
> I wonder why we would we use dist_autograd when there is no RPC in the forward pass? We probably should error out explicitly in this case and tell the application to use local autograd engine?

I agree with @mrshenli 's point. That would solve the concurrent cases in the PR summary.
@albanD @ezyang what do you guys think?

hczhu commented:
This PR doesn't work for the intended use case in #33951, where a remote node will have multiple parallelized backward passes on different autgrad graphs sharing the same `AccumulateGrad` nodes. We need another solution.
@albanD @mrshenli

hczhu commented:
To solve the concurrent backward pass issue, I'll try the delegation solution. More precisely, the following changes have to be made

```Cpp
class DistAccumulateGrad : public Node {
  DistAccumulateGrad(AccumulateGrad, DistAutogradContext);
  // Need to mark the method virtual in Node
  const std::vector<std::unique_ptr<FunctionPreHook>>& post_hooks() const override {
    return accumulateGrad_->post_hooks();
  }
  // Delegate pre_hooks() to 'AccumulateGrad'
  // Maybe delegate other methods as well

  // Override apply()
  
  // Keep the original AccumulateGrad intact and delegate necessary methods to it.
  AccumulateGrad accumulateGrad_;
};

```

`DistAccumulateGrad` will replace all `AccumulateGrad` nodes.

DDP's reducer keeps references to each output variable's `AccumulateGrad` and calls their `del_post_hook()`. That should just work.

If a backward pass is called on a graph without any RPC, throw an exception to tell the callsite to use the vallina engine.

@albanD @mrshenli @ezyang What do you guys think of the plan above?

albanD commented:
Is you point that ?
- We prevent mixing of the engines by forbidding using the distributed one without RPC
- Different distributed passes need to run the forward in the right context so they cannot share graph

I don't think these are true especially when you run with partial graphs:

```python
input = torch.rand(10, requires_grad=True)
common = model1(input)

# thread1
with dist_autograd.context() as context_id:
    loss = model2(common).sum()
    dist_autograd.backward(context_id, [loss])

# thread2
with dist_autograd.context() as context_id:
    loss = model2(common).sum()
    dist_autograd.backward(context_id, [loss])

# thread3
loss = model3(common).sum()
loss.backward()
```

hczhu commented:
@albanD Thanks for the feedback.
Do you agree on `DistAccumulateGrad` implementation? 

> Is you point that ?
> 
> * We prevent mixing of the engines by forbidding using the distributed one without RPC
> * Different distributed passes need to run the forward in the right context so they cannot share graph

Yes, that's right. I see it doesn't work for a shared subgraph outside `context()`.
I think we have to somehow "lock" all `AccumulateGrad` nodes of a graph in order to solve the concurrency issue. Something like:
```
// Traverse the graph and get all AccumulateGrad nodes as raw pointers.
// Lock all such node at once. Throw an exception if lock acquisition fails.
replaceAccumulateGradNodes(); 
// Do the backward pass
// Restore the graph
// Unlock all nodes
```
Does the snippet above look good?

albanD commented:
That would fix this issue but I'm not sure if it's a good solution overall.
@pritamdamania87 do you have an idea for a better way to do this?
Or is this limitation ok for the distributed engine?

pritamdamania87 commented:
> I wonder why we would we use dist_autograd when there is no RPC in the forward pass? We probably should error out explicitly in this case and tell the application to use local autograd engine?

This is a nice to have feature currently and works well since we just wrap the local autograd engine. Would be nice to keep this instead of throwing an exception. An example is when users are trying out a bunch of things and might remove RPCs to test something. What problem are we trying to solve by returning an error to the user in this case?

Regarding the concurrency issue can we do this:

1. Only support running one type of engine at a time. Either you only use the distributed engine or you only use the local engine, but not both at the same time.
2. Each backward pass in the distributed engine, first replaces outgoing edges to any AccumulateGrad node with a DistAccumulateGrad node. If there are multiple backward passes running in parallel, each backward pass has its own DistAccumulateGrad node for each leaf.
3. Since each backward pass has its own context and its own DistAccumulateGrad node, we don't need any locking since the graphs are completely independent and they are even writing to separate context ids.
4. If retain_graph=True, in that case replace the the outgoing edges to DistAccumulateGrad with the original AccumulateGrad node.

Would this work, or did I miss something in the discussion?

hczhu commented:
@pritamdamania87 

> 3. Since each backward pass has its own context and its own DistAccumulateGrad node, we don't need any locking since the graphs are completely independent and they are even writing to separate context ids.

It's true that each backward pass has its own `DistAccumulateGrad` nodes after replacing `AccumulateGrad` nodes. The concurrency issue comes from the graph modification itself. Each forward pass under different RPC contexts indeed has its own graph instance, but what's twisted here is each graph instance shares the same `AccumulateGrad` nodes, which are kind of tied to their variables. More specifically, multiple threads replacing the same `AccumulateGrad` node by moving out its post-hooks is a data race.

My latest proposal in the following comment does resolve that, but it's quite clumsy. @albanD has a better proposal without graph modification. I'll write it up together with my proposal for everyone's review.

https://github.com/pytorch/pytorch/pull/34501#issuecomment-599764785

pritamdamania87 commented:
>  More specifically, multiple threads replacing the same AccumulateGrad node by moving out its post-hooks is a data race.

Do we need to move out the post-hooks? Can post-hooks just be copied by making it a shared_ptr? Ideally, we shouldn't modify anything in AccumulateGrad in the approach I mentioned above.

hczhu commented:
> > More specifically, multiple threads replacing the same AccumulateGrad node by moving out its post-hooks is a data race.
> 
> Do we need to move out the post-hooks? Can post-hooks just be copied by making it a shared_ptr? Ideally, we shouldn't modify anything in AccumulateGrad in the approach I mentioned above.

We can do that, but that doesn't solve the shared sub-graph issue between multiple backward passes.

https://github.com/pytorch/pytorch/pull/34501#issuecomment-599688418

pritamdamania87 commented:
> We can do that, but that doesn't solve the shared sub-graph issue between multiple backward passes.

Why wouldn't this work in the case of shared sub-graph? If model1 doesn't have any RPC, this should just work since each backward pass would use their own DistAccumulateGrad node for `input`. If model1 has RPC, the example is not valid since that part needs to be wrapped inside a context.

pritamdamania87 reviewed (changes requested):
Requesting changes based on the discussion we had to use a gradCapturePostHook. @hczhu please re-request review once we have that approach ready. Thanks!

Commit:
[py][dist] Trigger pre/post hooks of AccumulateGrad nodes under distributed autograd

pritamdamania87 commented on torch/csrc/autograd/engine.h:
```diff
@@ -53,6 +53,11 @@ struct GraphTask {
   std::unordered_map<Node*, InputBuffer> not_ready_;
   std::unordered_map<Node*, int> dependencies_;
 
+  struct GraphTaskFunctionPreHook {
```
nit: Should we call this GradCapturePostHook as we mentioned it in the design doc?

hczhu replied:
Ok, I'll rename it.

hczhu replied:
See the reply here:
https://github.com/pytorch/pytorch/pull/34501/files#r397583014

pritamdamania87 commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -587,6 +587,9 @@ void Engine::evaluate_function(
     }
+    for (auto& hook : fn_info.hooks_) {
+      (*hook)(inputs.toVariables());
+    }
```
If this is a grad capture post hook, these hooks should be executed only if `fn_info.needed_` is false?

hczhu replied:
I want to make the hook execution orthogonal to `needed_`. The callsite has `neede_` info and can decide whether to install the hook accordingly if necessary.
Otherwise, a callsite may get surprised if the hook of a not-needed node is not executed in the future.

pritamdamania87 commented on torch/csrc/autograd/input_buffer.h:
```diff
@@ -33,7 +33,12 @@ struct InputBuffer {
+  Variable operator[](size_t pos) {
+    return buffer[pos];
+  }
+  const std::vector<Variable>& toVariables() const {
```
nit: getInputs() seems like a better name.

pritamdamania87 commented on torch/csrc/distributed/autograd/context/context.h:
```diff
@@ -71,6 +71,7 @@ class TORCH_API DistAutogradContext {
   friend class BackwardPassCleanupGuard;
   friend class DistEngine;
   friend class RecvRpcBackward;
+  friend class AccumulateGradPreHook;
```
nit: DistAccumulateGradCaptureHook

hczhu replied:
Actually the execution of this hook is not tied to output capture in the local engine. That means a hook can be installed and executed even if an `AccumulateGrad` node doesn't have `captures_` set  in`exec_info_`.

hczhu replied:
@albanD @pritamdamania87 
Should we stop capturing the grads for output nodes in dist_engine? My understanding is capturing the grads is for accumulating them into RPC context in the graph task callback. `graphTask->init_to_execute(dummyRoot, outputEdges);` call does the trick to tell the local engine to capture grads for `outputEdges`.

pritamdamania87 commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -24,6 +25,38 @@ static constexpr char* kEngineCPUQueueSize =
     "local_autograd_engine_cpu_queue_size";
 static constexpr char* kNumAutogradContexts = "num_autograd_contexts";
 
+struct AccumulateGradPreHook : GraphTask::GraphTaskFunctionPreHook {
```
Separate implementation in .h and .cpp file? Also add some docs to this class stating its purpose.

hczhu replied:
That's not necessary. This is an internal class dedicated to `DistEngine`.

hczhu replied:
Added comments.

pritamdamania87 commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -24,6 +25,38 @@ static constexpr char* kEngineCPUQueueSize =
+    // It is possible that the grad is not defined since a separate
+    // invocation of the autograd engine on the same node might actually
+    // compute this gradient. Also accumulate grads only for
+    // AccumulateGrad function.
```
"Also accumulate grads only for AccumulateGrad function". This part of the comment is not valid here?

hczhu replied:
deleted.

pritamdamania87 commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -175,6 +208,18 @@ void DistEngine::computeDependencies(
+              accumulateGradFn->shared_from_this()),
+          autogradContext));
+    }
+  }
```
We visit the AccumulateGrad nodes in the BFS above when we compute dependencies. Can we attach the post hooks at that point?

pritamdamania87 reviewed (changes requested):
Looks good overall, have some comments inline regarding code structure. In terms of testing can we update https://github.com/pytorch/pytorch/pull/33951 to have all the necessary assertions and probably stack that PR on top of this one to ensure we have good test coverage for this code?

Commit:
Addressed comments

pritamdamania87 commented:
cc @zzzwen

pritamdamania87 commented on torch/testing/_internal/distributed/rpc/dist_autograd_test.py:
```diff
@@ -2065,6 +2065,20 @@ def backward(ctx, grad):
+        self.hook_called_times = 0
+        def post_hook(m, grad_input, grad_output):
+            self.hook_called_times += 1
+        # Note: this actually doesn't install a post hook in AccumulateGrad.
```
This doesn't really test the codepaths we have modified in this PR right? We should test post hooks on AccumulateGrad itself. An example of how to add post hooks to AccumulateGrad:

```
import torch

count = 0
def post_hook(grad_input, grad_output):
    global count
    count += 1
    return grad_input

t = torch.rand(10, 10, requires_grad=True)
a = t + t
accumulate_grad = a.grad_fn.next_functions[0][0]
accumulate_grad.register_hook(post_hook)
a.sum().backward()
print(count)
```

Lets test both cases: 1) post hooks on non-leaf grad functions 2) post hooks on accumulate grad

hczhu replied:
> 1) post hooks on non-leaf grad functions

We can test that, but want to make sure we are on the same page. Post hooks on non-leaf functions (e.g. `a.grad_fn` here) already work without this PR.

> 2) post hooks on accumulate grad

I tried the example code. `t.grad_fn` is `None` in this example. Any suggestion?

@pritamdamania87

pritamdamania87 replied:
> We can test that, but want to make sure we are on the same page. Post hooks on non-leaf functions (e.g. a.grad_fn here) already work without this PR.

Yes, I meant to keep the current test.

> I tried the example code. t.grad_fn is None in this example. Any suggestion?

Yes, thats why I had to use `a.grad_fn.next_functions[0][0]` to get accumulate_grad. Seems like you would have to do something similar. I'm not sure why t.grad_fn is None though. cc @albanD Is this expected for accumulate grad?

albanD replied:
> Is this expected for accumulate grad?

Yes, this is expected. In particular for the leaf definition.
The AccumulateGrad Node is a trick we use to accumulate the gradient, but the Tensor is a leaf that requires grad (so has no grad_fn) as it was not created in a differentiable manner.

pritamdamania87 commented on torch/testing/_internal/distributed/rpc/dist_autograd_test.py:
```diff
@@ -2065,6 +2065,20 @@ def backward(ctx, grad):
+            self.hook_called_times += 1
+        # Note: this actually doesn't install a post hook in AccumulateGrad.
+        linear.register_backward_hook(post_hook)
+        linear.register_backward_hook(post_hook)
```
Can we test different post hooks here?

pritamdamania87 reviewed (changes requested):
Thanks for adding the unit test, requesting changes since I think we should test post hooks on AccumulateGrad itself.

Commit:
address comments

Commit:
Fix build

Commit:
Fix build

albanD reviewed (changes requested):
As discussed offline, the signature of the hook needs to be updated.
Ping me when it is ready for review.

Commit:
Change according to offline discussion.
1. Make the new hook a part of grad capture.
2. Change the hook signature to accept a grad.

albanD commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -23,6 +24,43 @@ using torch::autograd::variable_list;
+        autogradContext_(std::move(autogradContext)) {}
+
+  torch::Tensor operator()(const torch::Tensor& grad) override {
+    // It is possible that the grad is not defined since a separate
```
Not sure what that comment means?
undefinted gradient means it is a Tensor full of 0s. You might want to double check the posthook logic in the engine, but I this we do call the post hooks even when gradients are undefined.

pritamdamania87 replied:
Is there a reason we do this for the local engine? It seems like when the grad is not defined, we don't call pre-hooks and skip gradient accumulation, but still call post hooks? I'm fine with replicating the logic in the local engine, but it seems inconsistent that we only call post hooks.

hczhu replied:
I copy-pasted the comment from the code I deleted. This is to avoid accumulating an undefined grad into the RPC context. 
@pritamdamania87 This current implementation is to skip everything if the grad is not defined, which is consistent with calling pre-hooks.


https://github.com/pytorch/pytorch/blob/c7631716da203849d15d17231db6ee0304363471/torch/csrc/distributed/autograd/engine/dist_engine.cpp#L233-L237

albanD replied:
We do call both pre and post hooks all the time.
Only the special variable hooks called from within the AccumulateGrad are skipped when the grad is undefined.

pritamdamania87 replied:
I see, let's keep this consistent and still call the post hooks even if the grad is not defined.

hczhu replied:
Ok, will call post hooks even if the grad is undefined.

albanD commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -23,6 +24,43 @@ using torch::autograd::variable_list;
+      const variable_list kEmptyOuput;
+      const variable_list inputGrads = {grad};
+      for (const auto& hook : accumulateGrad_->post_hooks()) {
+        // Discard the return value.
```
There should be no return value here as the output is empty right?

hczhu replied:
The post hook signature has a return value, `variable_list`. It's possible that a subclass of the base pre hook returns an non-empty `variable_list`.

https://github.com/pytorch/pytorch/blob/8544591f5a312980c353e9ff2db8a357943a3db5/torch/csrc/autograd/function_hook.h#L19-L24

albanD replied:
Since the hook cannot change the number or size of the gradient, you should raise an error if the hook returns anything else than an empty variable_list right?

albanD commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -180,6 +218,22 @@ void DistEngine::computeDependencies(
+      if (!execInfo.captures_) {
+        continue;
+      }
+      if (auto accumulateGradFn = dynamic_cast<AccumulateGrad*>(fn)) {
```
Shouldn't this be an assert? 
This whole logic is going to do weird things if the node is not an accumulate grad (namely not capture the gradients properly).

hczhu replied:
Wil do that.

albanD commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -23,6 +24,43 @@ using torch::autograd::variable_list;
 
+// This hook does 2 things:
+//   1. Accumuate the gard to RPC context.
+//   2. Call post hooks of the original AccumulateGrad.
```
Is there a reason why you call the post hooks but not the pre-hooks here?

pritamdamania87 replied:
`autogradContext_->accumulateGrad` calls the pre-hooks, although it would be nice to move that part out into this function such that the structure looks like:

1) Call pre-hooks.
2) Accumulate Grad
3) Call post-hooks

We should also rename `AccumulateGrad::callHooks` to `AccumulateGrad::callPreHooks`

hczhu replied:
Should we address what @pritamdamania87 suggested in another PR?

albanD replied:
@pritamdamania87 I don't think we should rename "callHooks". It is not calling the pre-hooks but the hooks that are associated with the leaf Variable that is given.

pritamdamania87 replied:
I was referring to them as pre-hooks since they're a list of `FunctionPreHook`. Maybe we can rename it to `callVariableHooks` or something. @hczhu We can look at this in a separate PR as you suggested.

hczhu replied:
> Is there a reason why you call the post hooks but not the pre-hooks here?

I didn't think about that at all, since the goal of this hook is to trigger DDP post hooks.
@pritamdamania87 @albanD How do you think? Should we call pre hooks as well here?

albanD replied:
Yes, Pritam mentionned it in the other comment below.

albanD commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -630,7 +630,10 @@ void Engine::evaluate_function(
-      for (auto capture : *capture_vec) {
+      for (const auto& capture : *capture_vec) {
+        for (auto& hook : capture.hooks_) {
+          inputs[capture.input_idx_] = (*hook)(inputs[capture.input_idx_]);
```
What is the justification to modify the input to the next Node here?
Shouldn't this only modify the captured value?

hczhu replied:
It's a contract similar to what a chain of function pre-hooks does. 
Which returned value to capture if there are multiple hooks, if a hook only modifies the captured grad? Only the last one counts?

albanD replied:
For the captured value I agree: the returned value should be the one captured.
But for the rest of the graph and the following Nodes I'm not so sure.

albanD reviewed (commented):
Looks mostly ok, just a few questions.

pritamdamania87 commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -180,6 +218,22 @@ void DistEngine::computeDependencies(
     // Create a dummy GraphRoot and run init_to_execute with it.
     GraphRoot dummyRoot(edges, {});
     graphTask->init_to_execute(dummyRoot, outputEdges);
+    for (auto& pr : graphTask->exec_info_) {
```
nit: can we call this `mapEntry`

hczhu replied:
Will change it.

pritamdamania87 commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -180,6 +218,22 @@ void DistEngine::computeDependencies(
+        continue;
+      }
+      if (auto accumulateGradFn = dynamic_cast<AccumulateGrad*>(fn)) {
+        for (auto& capture : *execInfo.captures_) {
+          capture.hooks_.push_back(
+              std::make_unique<DistAccumulateGradCapturePreHook>(
+                  std::dynamic_pointer_cast<AccumulateGrad>(
+                      accumulateGradFn->shared_from_this()),
+                  autogradContext));
+        }
+      }
```
We used to have this logic in the BFS above (line 166 - 185) to avoid traversing exec_info again. Is there a reason we moved this logic here instead?

hczhu replied:
Yes, in the BFS the `exec_info` doesn't even have entries for `AccumulateGrad` nodes.  The call of `init_to_execute()` after the BFS adds those entries. I don't want to replicate the similar logic here.

pritamdamania87 reviewed (changes requested):
Thanks for updating the PR with the latest design! Overall structure looks good. I'm requesting changes for some inline comments and also since it looks like CI is broken.

Commit:
Addressed comments

Commit:
Addressed comments

Commit:
Fix build

hczhu commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -544,7 +544,7 @@ static void _trace_post_record(
   }
 
-  node->i_(attr::inplace, is_inplace);
+  node->i_(jit::attr::inplace, is_inplace);
```
Build is broken for me. Not sure it's just me. I had to change this.

pritamdamania87 replied:
Can you try a rebase to see if that fixes the problem? Is the internal build broken or the OSS one?

hczhu replied:
OSS one. The internal build is also broken. I sent out a diff to fix it.

hczhu replied:
rebasing doesn't solve the problem.

pritamdamania87 commented on torch/csrc/autograd/engine.h:
```diff
@@ -56,10 +57,21 @@ struct GraphTask {
       int input_idx_; // within Node inputs
       int output_idx_; // within the output vector of a GraphTask
+
+      // This hook will be executed before a grad is captured. The engine will
+      // capture the returned value.
+      struct GradCapturePreHook {
+        virtual ~GradCapturePreHook() = default;
+        virtual torch::Tensor operator()(const torch::Tensor& grad) = 0;
+      };
```
nit: Maybe call this GradCaptureHook since we're actually calling the hook after the grad is captured in captured_vars_ now. Probably need to fix the comment as well?

hczhu replied:
Will make the change.

pritamdamania87 commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -23,6 +23,45 @@ using torch::autograd::variable_list;
+// This hook does 2 things:
+//   1. Accumuate the gard to RPC context.
+//   2. Call post hooks of the original AccumulateGrad.
+struct DistAccumulateGradCapturePreHook
```
nit: If we're renaming GradCapturePreHook as I suggested above, we can rename this too.

pritamdamania87 reviewed (commented):
Looks good to me. I guess the only thing missing is calling the pre-hook in DistAccumulateGradCapturePreHook as @albanD suggested. I'll stamp the PR once that is added.

Commit:
Call pre-hooks

Commit:
Fix comments

hczhu commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -23,6 +23,57 @@ using torch::autograd::variable_list;
+    // It's intended that pre/post hooks are still called even if the grad is
+    // undenfined here.
+    if (!accumulateGrad_->pre_hooks().empty()) {
+      variable_list inputGrads = {grad};
+      for (const auto& hook : accumulateGrad_->pre_hooks()) {
+        inputGrads = (*hook)(inputGrads);
+        // A pre-hook should return a varaible list of the same size as the
+        // function's inputs.
+        TORCH_INTERNAL_ASSERT(inputGrads.size() == 1);
+      }
+      grad = inputGrads[0];
```
The input grad is modified by the pre-hooks. The intention is to keep the same semantic of pre-hooks. Let me if it shouldn't modify the input grad. @pritamdamania87  @albanD

albanD replied:
Yes the hooks modify the value that go to the next hook and then next steps.

albanD commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -23,6 +23,57 @@ using torch::autograd::variable_list;
+    }
+    // It's intended that pre/post hooks are still called even if the grad is
+    // undenfined here.
+    if (!accumulateGrad_->pre_hooks().empty()) {
```
The pre-hooks should be called before the grad accumulation:
- Pre-hook
- Accumulation
- Post-hook.

Commit:
Fix build

Commit:
Simplify code

albanD commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -23,6 +23,52 @@ using torch::autograd::variable_list;
+      inputGrads = (*hook)(inputGrads);
+      // A pre-hook should return a varaible list of the same size as the
+      // function's inputs.
+      TORCH_INTERNAL_ASSERT(inputGrads.size() == 1);
```
nit: in the engine, such check does not exist (not sure why) so it's not necessary here. If you want to keep it (which makes sense here since this is an AccumulateGrad Node), I don't think it should be an internal assert though. This is user code that can return anything. I would lean towards something like: `TORCH_CHECK(inputGrads.size() == 1, "Pre hook for AccumulateGrad returned an invalid number of elements, got ", inputGrad.size(), " but 1 was expected");`.

hczhu replied:
I'll remove those 2 checks.

Commit:
fix build

hczhu commented on torch/csrc/autograd/engine.h:
```diff
@@ -10,6 +10,7 @@
 #include <torch/csrc/autograd/functions/basic_ops.h>
 #include <torch/csrc/autograd/input_buffer.h>
 #include <torch/csrc/utils/future.h>
+#include <torch/types.h>
```
This file is gone. Should be `#include <ATen/Tensor.h>`

Commit:
Fix build & addressed comments

Commit:
Fix clang complaint

Commit:
Fix test

Commit:
Fix a bug

Commit:
Fix test

Commit:
fix build

hczhu commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -23,6 +23,53 @@ using torch::autograd::variable_list;
+    // invocation of the autograd engine on the same node might actually
+    // compute this gradient.
+    if (inputGrads[0].defined()) {
+      // Three internal references to 'inputGrads[0]' at this moment:
+      //   1. 'inputGrads[0]'
+      //   2. 'grad'
+      //   3. The captured grad in the callsite.
+      autogradContext_->accumulateGrad(
+          accumulateGrad_->variable,
+          inputGrads[0],
+          3 /* num_expected_refs */);
```
Hope my understanding for `num_expected_refs` is correct here. The test case for non-copying grad does pass.
@pritamdamania87

pritamdamania87 replied:
> 2. 'grad'

This should not bump the reference count of the Tensor since it is `const Tensor&` ?

> 3. The captured grad in the callsite.

Is this is referring to `captured_vars_` ?, if so lets mention `captured_vars_` here instead?

pritamdamania87 replied:
Where is the third reference coming from? Aren't there 2 references? One in `captured_vars_ `and one in `inputGrads`?

hczhu replied:
It's indeed 3. I printed `use_count()` in a test. Another one is `inputs` in the callsite. You are right that `grad` doesn't bump ref count. 

```
InputBuffer& inputs) {
```
Anyway, the ref count is quite fragile here. A callsite in the future can have an external reference (no other internal references) to the grad and this operator will incorrectly move(steal) a sparse tensor.

albanD replied:
nit: maybe add a comment here explaining where these come from so that we won't have to hunt them down next time.

pritamdamania87 commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -23,6 +23,53 @@ using torch::autograd::variable_list;
 static constexpr char* kNumAutogradContexts = "num_autograd_contexts";
 
+// This hook does 3 things:
+//   1. Accumuate the gard to RPC context.
```
nit: grad and reorder the steps here? prehook, accumulate and then post hook?

pritamdamania87 reviewed (commented):
Do we need still the changes in torch/csrc/autograd/python_function.cpp and torch/csrc/cuda/Module.cpp? I'm able to compile on master without those changes.

hczhu commented on torch/csrc/cuda/Module.cpp:
```diff
@@ -35,7 +35,7 @@ static bool in_bad_fork = false;  // True for children forked after cuda init
 static void forked_child() {
   in_bad_fork = true;
-  utils::set_run_yet_variable_to_false();
+  torch::utils::set_run_yet_variable_to_false();
```
Yes, otherwise build will fail. It's the right thing to do as well. Search for other instances of `torch::utils` below in this file.

albanD replied:
Why is this needed here but wasn't in master?

hczhu replied:
Resolved offline. Will keep this change.

albanD commented on torch/csrc/autograd/engine.h:
```diff
@@ -64,10 +65,25 @@ struct GraphTask {
+        virtual ~GradCaptureHook() = default;
+        virtual at::Tensor operator()(const at::Tensor& grad) = 0;
+      };
+      // The hooks will be called one by one in the order. The input grad of a
```
Which order? Is it the same order as they were added?

hczhu replied:
yes, will update the comment.

albanD commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -544,7 +544,7 @@ static void _trace_post_record(
   }
 
-  node->i_(attr::inplace, is_inplace);
+  node->i_(jit::attr::inplace, is_inplace);
```
Is this still needed??

albanD reviewed (approved):
LGTM except the extra unrelated changes

Commit:
Addressed comments

Commit:
revert unneeded changes

wanchaol commented:
@hczhu it seems that the PR messed up when merging with the master, can you rebase again?
